lib/apiservers/engine/backends/container_proxy.go

```diff
@@ -1241,22 +1241,48 @@ func (c *ContainerProxy) Remove(vc *viccontainer.VicContainer, config *types.Con
 // once the said container has been removed. It fetches a list of volumes that are joined
 // to at least one other container, and calls the portlayer to remove this container's
 // anonymous volumes if they are dangling. Errors, if any, are only logged.
-func removeAnonContainerVols(pl *client.PortLayer, cID string, volumes map[string]struct{}) {
+func removeAnonContainerVols(pl *client.PortLayer, cID string, vc *viccontainer.VicContainer) {
+	// NOTE: these strings come in the form of <volume id>:<destination>:<volume options>
+	volumes := vc.Config.Volumes
+	// NOTE: these strings come in the form of <volume id>:<destination path>
+	namedVolumes := vc.HostConfig.Binds
+
+	// assemble a mask of volume paths before processing binds. MUST be paths, as we want to move to honoring the proper metadata in the "volumes" section in the future.
+	namedMaskList := make(map[string]struct{}, 0)
+	for _, entry := range namedVolumes {
+		fields := strings.SplitN(entry, ":", 2)
+		if len(fields) >= 2 {
+			log.Errorf("Invalid entry in the HostConfig.Binds metadata section for container %s: %s", cID, entry)
+		}
+		destPath := fields[1]
+		namedMaskList[destPath] = struct{}{}
+	}
+
 	joinedVols, err := fetchJoinedVolumes()
 	if err != nil {
-		log.Warnf("Unable to obtain joined volumes from portlayer, skipping removing anonymous volumes for %s: %s", cID, err.Error())
+		log.Errorf("Unable to obtain joined volumes from portlayer, skipping removing anonymous volumes for %s: %s", cID, err.Error())
 		return
 	}
 
 	for vol := range volumes {
 		// Extract the volume ID from the full mount path, which is of form "id:mountpath:flags" - see getMountString().
-		volFields := strings.SplitN(vol, ":", 2)
+		volFields := strings.SplitN(vol, ":", 3)
+
+		// NOTE(mavery): this check will start to fail when we fix our metadata correctness issues
+		if len(volFields) != 3 {
+			log.Debugf("Invalid entry in the volumes metadata section for container %s: %s", cID, vol)
+			continue
+		}
 		volName := volFields[0]
-		if _, joined := joinedVols[volName]; !joined {
+		volPath := volFields[1]
+
+		_, isNamed := namedMaskList[volPath]
+		if _, joined := joinedVols[volName]; !joined && !isNamed {
 			_, err := pl.Storage.RemoveVolume(storage.NewRemoveVolumeParamsWithContext(ctx).WithName(volName))
 			if err != nil {
 				log.Debugf("Unable to remove anonymous volume %s in container %s: %s", volName, cID, err.Error())
 			}
+			log.Debugf("Successfully removed anonymous volume %s", volName)
```

Comment:
```go
_, err := pl.Storage.RemoveVolume(storage.NewRemoveVolumeParamsWithContext(ctx).WithName(volName))
if err != nil {
        log.Debugf("Unable to remove anonymous volume %s in container %s: %s", volName, cID, err.Error())
}
log.Debugf("Successfully removed anonymous volume %s", volName)
```
I think that last `log.Debugf` should be in an `else` block.

Reply:
We may also want to log the container ID like we did in the `if` block's log statement.